### PR TITLE
Rename/port net_ipc virtio backend to net_vpnkit

### PR DESCRIPTION
**Pull request by ijc25**
When this was developed vpnkit didn't yet have a name, now that it does rename
it to fit and add the frontend protocol inline instead of expecting it
externally (and simplify things along the way).

Signed-off-by: Ian Campbell <ian.campbell@docker.com>

### CODE REVIEW

File: src/pci_virtio_net_vpnkit.c [hunk 1]
```diff
@@ -78,15 +78,37 @@
...
+struct msg_ethernet {
+	uint8_t command; /* CMD_ETHERNET */
+	char uuid[36];
+} __packed;
+
+struct vif_info {
+	uint16_t mtu;
+	uint16_t max_packet_size;
+	uint8_t mac[6];
+} /*__packed*/;
```
Comment (avsm) [inline src/pci_virtio_net_vpnkit.c]: Why's this commented out?

File: src/pci_virtio_net_vpnkit.c [hunk 2]
```diff
@@ -877,17 +969,10 @@ pci_vtnet_ping_ctlq(void *vsc, struct vqueue_info *vq)
...
 	int mac_provided;
 	pthread_t sthrd;
 
-	if (!aslInitDone) {
-		aslInit("Docker", "com.docker.hyperkit");
-		aslLog(ASL_LEVEL_NOTICE, "Starting hyperkit");
-		aslInitDone = 1;
-	}
-
 	sc = calloc(1, sizeof(struct pci_vtnet_softc));
```
Comment (avsm) [inline src/pci_virtio_net_vpnkit.c]: appears to be a `calloc` NULL check missing here.

File: src/pci_virtio_net_vpnkit.c (hunk 1 above)
Comment (ijc25) [inline src/pci_virtio_net_vpnkit.c]: Clang complained (with an error) that the struct was already naturally packed and so therefore the annotation was unnecessary. I left it in mainly for its documentary effects (personally I think the error was actively harmful...)

File: src/pci_virtio_net_vpnkit.c (hunk 2 above)
Comment (ijc25) [inline src/pci_virtio_net_vpnkit.c]: That code isn't changed here, so that would need to be a separate PR.

File: src/pci_virtio_net_vpnkit.c [hunk 3]
```diff
@@ -196,13 +218,138 @@ static struct virtio_consts vtnet_vi_consts = {
...
 	int fd;
 	struct vif_info vif;
 };
 
 #pragma clang diagnostic pop
 
+static int really_read(int fd, uint8_t *buffer, size_t total)
+{
+	size_t remaining = total;
+        ssize_t n;
```
Comment (rneugeba) [inline src/pci_virtio_net_vpnkit.c]: nitpick: indentation

File: src/pci_virtio_net_vpnkit.c [hunk 4]
```diff
@@ -196,13 +218,138 @@ static struct virtio_consts vtnet_vi_consts = {
...
+
+err:
+	/* On error: stop reading from the socket and trigger a clean shutdown */
+	shutdown(fd, SHUT_RD);
+	return -1;
+}
+
+static int really_write(int fd, uint8_t *buffer, size_t total) {
+        size_t remaining = total;
+	ssize_t n;
```
Comment (rneugeba) [inline src/pci_virtio_net_vpnkit.c]: nitpick: indentation

File: src/pci_virtio_net_vpnkit.c [hunk 5]
```diff
@@ -196,13 +218,138 @@ static struct virtio_consts vtnet_vi_consts = {
...
+	memcpy(cmd_ethernet.uuid, uuid, sizeof(cmd_ethernet.uuid));
+
+	if (really_write(fd, (uint8_t*)&cmd_ethernet, sizeof(cmd_ethernet)) < 0) {
+		fprintf(stderr, "virtio-net-vpnkit: failed to write ethernet cmd\n");
+		return -1;
+	}
+
+	if (really_read(fd, (uint8_t*)vif, sizeof(*vif)) < 0) {
+		fprintf(stderr, "virtio-net-vpnkit: failed to vif info\n");
+		return -1;
```
Comment (rneugeba) [inline src/pci_virtio_net_vpnkit.c]: failed to "read" vif info?

**Comment by rneugeba** [comment]
Again, not too familiar with the code, but apart from some minor nitpicks, LGTM

**Comment by ijc25** [comment]
Thank's Rolf, I fixed your nitpicks and repushed.

File: src/pci_virtio_net_vpnkit.c (hunk 1 above)
Comment (ijc25) [inline src/pci_virtio_net_vpnkit.c]: FYI:
```
cc src/pci_virtio_net_vpnkit.c
src/pci_virtio_net_vpnkit.c:106:11: error: packed attribute is unnecessary for 'mtu' [-Werror,-Wpacked]
        uint16_t mtu;
                 ^
src/pci_virtio_net_vpnkit.c:107:11: error: packed attribute is unnecessary for 'max_packet_size' [-Werror,-Wpacked]
        uint16_t max_packet_size;
                 ^
src/pci_virtio_net_vpnkit.c:105:8: error: packed attribute is unnecessary for 'struct vif_info' [-Werror,-Wpacked]
struct vif_info {
       ^
3 errors generated.
```

Comment (ijc25) [inline src/pci_virtio_net_vpnkit.c]: In greping for where `-Wpacked` came from I found several instances of
```
#pragma clang diagnostic ignored "-Wpacked"
```
in the code, I think I shall do the same here.

**Comment by djs55** [comment]
LGTM
